### PR TITLE
Yongbing chen/human in the loop

### PR DESCRIPTION
**Pull request by yongbing-chen**

### CODE REVIEW

File: src/react/components/pages/predict/uploadToTrainingSetView.tsx [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        this.setState({ isLoading: true });
+        if (this.props.onConfirm) {
+            await this.props.onConfirm();
+        }
+        this.close();
+    }
+    render() {
+        const dark: ICustomizations = {
+            settings: {
+                theme: getDarkGreyTheme(),
```
Comment (alex-krasn) [inline src/react/components/pages/predict/uploadToTrainingSetView.tsx:43]: you need to apply getDefaultDarkTheme() - otherwise the spinner looks like white circle (not progress visible)
```suggestion
                theme: getDefaultDarkTheme(),
```

Screen Shot 2020-08-27 at 9 26 52 AM

**Comment by alex-krasn** [comment]
I'm getting this error every time I try push "Edit & upload to train set" button: 
image

**Comment by stew-ro** [comment]
> I'm getting this error every time I try push "Edit & upload to train set" button:
> image

Also after closing this error, FOTT is stuck in a state which the user can't get out of. They then have to refresh the page

**Comment by alex-krasn** [comment]
> > I'm getting this error every time I try push "Edit & upload to train set" button:
> > image
> 
> Also after closing this error, FOTT is stuck in a state which the user can't get out of. They then have to refresh the page

yep

**Comment by alex-krasn** [comment]
The fix - solve first issue.

**Comment by alex-krasn** [comment]
now getting this error on web-version, on upload and if click on this document in editor also:
Screen Shot 2020-08-28 at 11 16 52 AM

**Comment by alex-krasn** [comment]
And on electron version:
1. 
image
2. Labels are all gray and when switch document retain labels from prev file:
image
image
o

**Comment by alex-krasn** [comment]
still getting this error on web:
image

**Review by stew-ro — CHANGES_REQUESTED** [review]
> 1)  "Edit & upload to training set" button is enabled when there's no fields

Fixed: allow upload the file when no fileds find.
image

> 2) Returning to the analyze page, after uploading to training set, shows "no field can be extracted"

Fixed
image

> 3)  Document files are overwritten without warning if the training set already has the document

Fixed: will popup a modal notify the user

> 4) Error for documents analyzed by URL

Fixed
image
image

**Comment by alex-krasn** [comment]
I'm getting error when running 'auto labeling' on other document in a project (not freshly "uploaded")
image

File: src/react/components/pages/editorPage/canvas.tsx [hunk 2]
```diff
@@ -293,32 +300,40 @@ export default class Canvas extends React.Component<ICanvasProps, ICanvasState>
...
                         Page {this.state.currentPage} of {this.state.numPages}
                     </p>
                 }
-                { this.state.ocrStatus !== OcrStatus.done &&
+                {this.state.ocrStatus !== OcrStatus.done &&
                     <div className="canvas-ocr-loading">
                         <div className="canvas-ocr-loading-spinner">
                             <Label className="p-0" ></Label>
-                            <Spinner size={SpinnerSize.large} label="Running OCR..." ariaLive="assertive" labelPosition="right"/>
+                            <Spinner size={SpinnerSize.large} label="Running OCR..." ariaLive="assertive" labelPosition="right" />
```
Comment (stew-ro) [inline src/react/components/pages/editorPage/canvas.tsx:328]: The only difference is the spinner label value. Factor out redundant/duplicate code

File: src/react/components/pages/editorPage/canvas.tsx [hunk 3]
```diff
@@ -1807,12 +1846,17 @@ export default class Canvas extends React.Component<ICanvasProps, ICanvasState>
...
+            if (labelType) {
+                regionCategory = labelType;
+            } else if (tag.type === FieldType.SelectionMark) {
+                regionCategory = FeatureCategory.Checkbox;
+            } else {
+                regionCategory = FeatureCategory.Text;
+            }
+        }
+        catch (err) {
+            console.error(err);
```
Comment (stew-ro) [inline src/react/components/pages/editorPage/canvas.tsx:1859]: there should be no error here. We fixed the tag type undefined error with #557

Comment (stew-ro) [reply src/react/components/pages/editorPage/canvas.tsx:1859]: so you don't need a try catch

File: src/services/predictService.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        const endTime = Number(new Date()) + (timeout || 10000);
+        interval = interval || 100;
+
+        const checkSucceeded = (resolve, reject) => {
+            const ajax = func();
+            ajax.then((response) => {
+                if (response.data.status.toLowerCase() === constants.statusCodeSucceeded) {
+                    resolve(response.data);
+                    // prediction response from API
+                    console.log("raw data", JSON.parse(response.request.response));
```
Comment (stew-ro) [inline src/services/predictService.ts:63]: remove console log

File: src/services/predictService.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+import ServiceHelper from "./serviceHelper";
+
+export enum AutoLabelingStatus {
+    none,
+    running,
+    done
+}
+export class PredictService {
+
+    constructor(private project: IProject) {
```
Comment (stew-ro) [inline src/services/predictService.ts:15]: remove constructor if it doesn't initialize anything

File: src/services/predictService.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    }
+    public async getPrediction(fileUrl: string): Promise<any> {
```
Comment (stew-ro) [inline src/services/predictService.ts:17]: should be static

File: src/common/localization/en-us.ts [hunk 7]
```diff
@@ -202,6 +202,14 @@ export const english: IAppStrings = {
         downloadScript: "Analyze with python script",
         defaultLocalFileInput: "Browse for a file...",
         defaultURLInput: "Paste or type URL...",
+        editAndUploadToTrainingSet: "Edit & upload to training set",
+        editAndUploadToTrainingSetNotify: "by clicking on this button, this form will be added to the Azure Storage Blob for this project, where you can edit these labels.",
+        editAndUploadToTrainingSetNotify2: "We are adding this file to your training set, where you could edit the labels and re-train the model.",
```
Comment (stew-ro) [inline src/common/localization/en-us.ts:207]: "can" not "could"

could is past tense

File: src/common/localization/en-us.ts [hunk 8]
```diff
@@ -202,6 +202,14 @@ export const english: IAppStrings = {
         downloadScript: "Analyze with python script",
         defaultLocalFileInput: "Browse for a file...",
         defaultURLInput: "Paste or type URL...",
+        editAndUploadToTrainingSet: "Edit & upload to training set",
+        editAndUploadToTrainingSetNotify: "by clicking on this button, this form will be added to the Azure Storage Blob for this project, where you can edit these labels.",
```
Comment (stew-ro) [inline src/common/localization/en-us.ts:206]: We support local file system as a connection, not just Azure storage. So this message should reflect that

File: src/services/predictService.ts (hunk 5 above)
Comment (yongbing-chen) [reply src/services/predictService.ts:15]: the constructor will declare private property named project

File: src/react/components/pages/predict/uploadToTrainingSetView.tsx (hunk 1 above)
Comment (yongbing-chen) [reply src/react/components/pages/predict/uploadToTrainingSetView.tsx:45]: fixed

**Comment by yongbing-chen** [comment]
> about electron version
the electron version is no guaranteed now, I am focus on the browser first.

**Comment by yongbing-chen** [comment]
> 1. tton is enabled when there's no fields



> 1. "Edit & upload to training set" button is enabled when there's no fields

Not an error, it is allowed upload the file when not fields find.

File: src/react/components/pages/editorPage/canvas.tsx (hunk 3 above)
Comment (yongbing-chen) [reply src/react/components/pages/editorPage/canvas.tsx]: removed and push to github

File: src/services/predictService.ts (hunk 4 above)
Comment (yongbing-chen) [reply src/services/predictService.ts:63]: removed

File: src/services/predictService.ts (hunk 5 above)
Comment (yongbing-chen) [reply src/services/predictService.ts:15]: the constructor is is required, it initialize the project proerty, the project is used in methods in PredictService

File: src/common/localization/en-us.ts (hunk 7 above)
Comment (yongbing-chen) [reply src/common/localization/en-us.ts:207]: fixed

File: src/common/localization/en-us.ts (hunk 8 above)
Comment (yongbing-chen) [reply src/common/localization/en-us.ts:206]: updated to: "by clicking on this button, this form will be added to this project, where you can edit these labels."

**Review by stew-ro — APPROVED** [review]

**Review by alex-krasn — APPROVED** [review]

**Review by kunzms — APPROVED** [review]
